File: src/harmonic_shifter/core/frequency_shifter.py
"""
Frequency shifting in the spectral domain.

This module implements linear frequency shifting by reassigning FFT bins.
"""

from typing import Tuple

import numpy as np


class FrequencyShifter:
    """
    Frequency shifter using spectral bin reassignment.

    Shifts all frequencies by a fixed Hz amount in the frequency domain.
    """

    def __init__(self, sample_rate: int, fft_size: int):
        """
        Initialize frequency shifter.

        Args:
            sample_rate: Audio sample rate in Hz
            fft_size: FFT size used for analysis

        Example:
            >>> shifter = FrequencyShifter(44100, 4096)
        """
        self.sample_rate = sample_rate
        self.fft_size = fft_size
        self.bin_resolution = sample_rate / fft_size
        self.n_bins = fft_size // 2 + 1

    def shift(
        self,
        magnitude: np.ndarray,
        phase: np.ndarray,
        shift_hz: float
    ) -> Tuple[np.ndarray, np.ndarray]:
        """
        Shift all frequencies by shift_hz in the spectral domain.

        Args:
            magnitude: (n_frames, n_bins) magnitude spectrum
            phase: (n_frames, n_bins) phase spectrum in radians
            shift_hz: Amount to shift in Hz (can be negative)

        Returns:
            Tuple of (shifted_magnitude, shifted_phase)

        Example:
            >>> mag, phase = stft(signal)
            >>> shifted_mag, shifted_phase = shifter.shift(mag, phase, 100)
        """
        n_frames, n_bins = magnitude.shape

        # Calculate bin shift
        bin_shift = int(round(shift_hz / self.bin_resolution))

        # Initialize output arrays
        shifted_magnitude = np.zeros_like(magnitude)
        shifted_phase = np.zeros_like(phase)

        # Shift each bin
        for k in range(n_bins):
            # Calculate target bin
            k_new = k + bin_shift

            # Check if target bin is within valid range
            if 0 <= k_new < n_bins:
                # Copy magnitude and phase to new bin
                shifted_magnitude[:, k_new] = magnitude[:, k]
                shifted_phase[:, k_new] = phase[:, k]

        return shifted_magnitude, shifted_phase

    def get_shifted_frequencies(self, shift_hz: float) -> np.ndarray:
        """
        Get shifted frequency values for each bin.

        Args:
            shift_hz: Frequency shift amount in Hz

        Returns:
            Array of shifted frequencies for each bin

        Example:
            >>> shifted_freqs = shifter.get_shifted_frequencies(100)
        """
        original_freqs = np.fft.rfftfreq(self.fft_size, 1.0 / self.sample_rate)
        return original_freqs + shift_hz

File: plugin/src/dsp/FrequencyShifter.cpp
#include "FrequencyShifter.h"
#include <algorithm>
#include <cmath>

namespace fshift
{

FrequencyShifter::FrequencyShifter(double sampleRate, int fftSize)
    : sampleRate(sampleRate),
      fftSize(fftSize),
      numBins(fftSize / 2 + 1),
      binResolution(static_cast<float>(sampleRate) / static_cast<float>(fftSize))
{
    // Pre-compute original frequencies for each bin
    originalFrequencies.resize(numBins);
    for (int i = 0; i < numBins; ++i)
    {
        originalFrequencies[i] = static_cast<float>(i) * binResolution;
    }
}

float FrequencyShifter::frequencyToBin(float frequencyHz) const
{
    return frequencyHz / binResolution;
}

float FrequencyShifter::binToFrequency(int bin) const
{
    return static_cast<float>(bin) * binResolution;
}

std::vector<float> FrequencyShifter::getShiftedFrequencies(float shiftHz) const
{
    std::vector<float> shifted(numBins);
    for (int i = 0; i < numBins; ++i)
    {
        shifted[i] = originalFrequencies[i] + shiftHz;
    }
    return shifted;
}

std::pair<std::vector<float>, std::vector<float>> FrequencyShifter::shift(
    const std::vector<float>& magnitude,
    const std::vector<float>& phase,
    float shiftHz)
{
    // Initialize output arrays with zeros
    std::vector<float> shiftedMagnitude(numBins, 0.0f);
    std::vector<float> shiftedPhase(numBins, 0.0f);

    // Calculate bin shift
    int binShift = static_cast<int>(std::round(shiftHz / binResolution));

    // Shift each bin
    for (int k = 0; k < numBins; ++k)
    {
        // Calculate target bin
        int kNew = k + binShift;

        // Check if target bin is within valid range
        if (kNew >= 0 && kNew < numBins)
        {
            // Copy magnitude and phase to new bin
            shiftedMagnitude[kNew] = magnitude[k];
            shiftedPhase[kNew] = phase[k];
        }
        // If target bin is outside range, energy is discarded (anti-aliasing)
    }

    return { shiftedMagnitude, shiftedPhase };
}

} // namespace fshift
